src/capabilities.rs

```diff
@@ -527,80 +580,118 @@ pub fn server_has_capability(ctx: &Context, feature: &'static str) -> bool {
+        }
+        probe_feature(entry, &mut features, CAPABILITY_WORKSPACE_SYMBOL);
+        probe_feature(entry, &mut features, CAPABILITY_FORMATTING);
+        probe_feature(entry, &mut features, CAPABILITY_RANGE_FORMATTING);
+        probe_feature(entry, &mut features, CAPABILITY_RENAME);
+        probe_feature(entry, &mut features, CAPABILITY_CODE_ACTIONS);
+        probe_feature(entry, &mut features, CAPABILITY_CODE_ACTIONS_RESOLVE);
+        probe_feature(entry, &mut features, CAPABILITY_CODE_LENS);
+        probe_feature(entry, &mut features, CAPABILITY_CALL_HIERARCHY);
+        features
+            .entry("lsp-diagnostics".to_string())
+            .or_default()
+            .push(server_name);
+        probe_feature(entry, &mut features, CAPABILITY_INLAY_HINTS);
 
-    if let Some(ref provider) = server_capabilities.execute_command_provider {
-        features.push(format!(
-            "lsp-execute-command:  commands: [{}]",
-            provider.commands.iter().join(", ")
-        ))
-    }
+        // NOTE controller should park request for capabilities until they are available thus it should
+        // be safe to unwrap here (otherwise something unexpectedly wrong and it's better to panic)
+        let server_capabilities = server_settings.capabilities.as_ref().unwrap();
 
-    if let Some(ref provider) = server_capabilities.semantic_tokens_provider {
-        let legend = match provider {
-            SemanticTokensServerCapabilities::SemanticTokensOptions(options) => &options.legend,
-            SemanticTokensServerCapabilities::SemanticTokensRegistrationOptions(regopts) => {
-                &regopts.semantic_tokens_options.legend
-            }
-        };
+        if let Some(ref provider) = server_capabilities.execute_command_provider {
+            features
+                .entry(format!(
+                    "lsp-execute-command: commands: [{}]",
+                    provider.commands.iter().join(", ")
+                ))
+                .or_default()
+                .push(server_name);
+        }
 
-        features.push(format!(
-            "lsp-semantic-tokens:     types: [{}]",
-            legend
-                .token_types
-                .iter()
-                .map(SemanticTokenType::as_str)
-                .join(", ")
-        ));
-        features.push(format!(
-            "lsp-semantic-tokens: modifiers: [{}]",
-            legend
-                .token_modifiers
-                .iter()
-                .map(SemanticTokenModifier::as_str)
-                .join(", ")
-        ));
+        if let Some(ref provider) = server_capabilities.semantic_tokens_provider {
+            let legend = match provider {
+                SemanticTokensServerCapabilities::SemanticTokensOptions(options) => &options.legend,
+                SemanticTokensServerCapabilities::SemanticTokensRegistrationOptions(regopts) => {
+                    &regopts.semantic_tokens_options.legend
+                }
+            };
+
```

Comment:
Using the first server's offset encoding is wrong.
I noticed you correct this in a later commit.
But this has too much potential for confusion that I will continue reading the sum of all changes rather than the individual commits.
If you want to fix this structure (make sure that every commit passes all tests and doesn't have known flaws) that would help but else I can do this, no problem.
I appreciate the amount of work that has gone into this, it's a great exercise.